### PR TITLE
Fix inconsistent Timer interface

### LINKED ISSUE #4457: Timer interface inconsistent: Timer::get_lap_time returns wall time
While working on #4456 I ran into a little disagreement with the current timer interface:

```
class Timer {
  /**
   * Access to the current CPU time without disturbing time measurement. The
   * elapsed time is returned in units of seconds.
   */
  double operator() () const;

  /**
   * Access to the current wall time without disturbing time measurement. The
   * elapsed time is returned in units of seconds.
   */
  double wall_time () const;

  /**
   * Return the last lap time; the time taken between the last start()/stop()
   * call.
   */
  double get_lap_time () const;
}
```

 - Improve wording: "without distubing time measurement" - that's actually not true :-)
 - inconsistent naming: `wall_time` versus `get_lap_time`
 - `get_lap_time` returns wall time

At the very least `get_lap_time` should CPU time consistent with `operator()`.

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #4457.

### CODE REVIEW

File: include/deal.II/base/timer.h [hunk 1]
```diff
@@ -732,18 +783,37 @@ Timer::get_data() const
...
 {
-  unsigned int my_id = dealii::Utilities::MPI::this_mpi_process(mpi_communicator);
-  if (my_id==0)
-    stream << mpi_data.max << " wall,"
-           << " max @" << mpi_data.max_index
-           << ", min=" << mpi_data.min << " @" << mpi_data.min_index
-           << ", avg=" << mpi_data.avg
-           << std::endl;
+  const Utilities::MPI::MinMaxAvg statistic = get_data();
+  stream << statistic.max << " wall,"
```
Comment (Rombur) [inline include/deal.II/base/timer.h]: Is there a reason why we do the output on every rank now?

File: source/base/timer.cc [hunk 2]
```diff
@@ -56,8 +57,10 @@ Timer::Timer()
 #endif
 {
 #ifdef DEAL_II_WITH_MPI
-  mpi_data.sum = mpi_data.min = mpi_data.max = mpi_data.avg = numbers::signaling_nan<double>();
-  mpi_data.min_index = mpi_data.max_index = numbers::invalid_unsigned_int;
+  mpi_data.sum = mpi_data.min = mpi_data.max = mpi_data.avg = 0.;
```
Comment (Rombur) [inline source/base/timer.cc]: Why did you change the values?

File: source/base/timer.cc [hunk 3]
```diff
@@ -77,13 +80,16 @@ Timer::Timer(MPI_Comm mpi_communicator,
...
   last_lap_time (0.),
+  last_lap_cpu_time (0.),
   running (false),
   mpi_communicator (mpi_communicator),
   sync_wall_time(sync_wall_time_)
 {
 #ifdef DEAL_II_WITH_MPI
-  mpi_data.sum = mpi_data.min = mpi_data.max = mpi_data.avg = numbers::signaling_nan<double>();
-  mpi_data.min_index = mpi_data.max_index = numbers::invalid_unsigned_int;
+  mpi_data.sum = mpi_data.min = mpi_data.max = mpi_data.avg = 0.;
```
Comment (Rombur) [inline source/base/timer.cc]: same

File: tests/base/timer_04.cc [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+  Assert(t.wall_time() > 0., ExcInternalError());
+  Assert(t.cpu_time() > 0., ExcInternalError());
+  Assert(t.last_wall_time() == 0., ExcInternalError());
+  Assert(t.last_cpu_time() == 0, ExcInternalError());
+
+  burn(5000);
+  Assert(t.stop() > 0., ExcInternalError());
+
+  Assert(t.last_wall_time() > 0., ExcInternalError());
```
Comment (Rombur) [inline tests/base/timer_04.cc]: Maybe check that the value is greater than the previous `last_wall_time`. Same for `cpu_time`.

File: tests/base/timer_04.cc [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+int main ()
+{
+  initlog();
+  deallog.threshold_double(1.e-10);
+
+  Timer t;
+  burn(50);
+
+  Assert(t.wall_time() > 0., ExcInternalError());
```
Comment (Rombur) [inline tests/base/timer_04.cc]: `Assert` does nothing in `Release` so this test only works in `Debug`. Use `AssertThrow` instead.

File: tests/base/timer_05.cc [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+// the same as timer_04.cc but this time with MPI.
```
Comment (Rombur) [inline tests/base/timer_05.cc]: I have the same comment as the previous test.

File: include/deal.II/base/timer.h (hunk 1 above)
Comment (masterleinad) [inline include/deal.II/base/timer.h]: As we usually use a ConditionalOstream object for dealing with output when using MPI it should be in the responsibility of the passed stream to decide what to do with the output. Do you agree?

File: source/base/timer.cc (hunk 2 above)
Comment (masterleinad) [inline source/base/timer.cc]: We initialize all the other data with zeros as well. In particular, this holds for `last_lap_time` which also stores information about the last lap. So I decided for consistency to also initialize all the other variables with zero.

**Comment by drwells** [comment]
If I understand this correctly, there are a few small incompatible changes in this patch. Would you mind summarizing them in another changelog entry?

File: source/base/timer.cc (hunk 2 above)
Comment (Rombur) [inline source/base/timer.cc]: That's fine. I am just wondering if we had a reason to use invalid value to initialize these variable.

File: include/deal.II/base/timer.h (hunk 1 above)
Comment (Rombur) [inline include/deal.II/base/timer.h]: Yeah it makes sense.

**Comment by Rombur** [comment]
LGTM.

**Comment by masterleinad** [comment]
@drwells Here you go!

File: source/base/timer.cc [hunk 7]
```diff
@@ -175,51 +181,48 @@ double Timer::stop ()
...
       last_lap_time = windows::wall_clock() - start_wall_time;
-      cumulative_time += windows::cpu_clock() - start_time;
+      last_lap_cpu_time = windows::cpu_clock() - start_time;
 #else
 #  error Unsupported platform. Porting not finished.
 #endif
 
 #ifdef DEAL_II_WITH_MPI
+      this->mpi_data = Utilities::MPI::min_max_avg (last_lap_time,
+                                                    mpi_communicator);
```
Comment (kronbichler) [inline source/base/timer.cc]: Why do you move the global communication out of `sync_wall_time`? Isn't this going to disturb the numbers by the MPI network? I agree that in order to get the overall wall time, one needs to communicate, but isn't this supposed to be done through the `sync_wall_time` passed to the respective constructor? (I know that it will often not matter when using `MPI_COMM_SELF`.)

File: source/base/timer.cc (hunk 2 above)
Comment (bangerth) [inline source/base/timer.cc]: At least in the signaling NaN case, the reason is always the same: if you *use* a signaling NaN, you will get an exception. That makes sense, because we initialize these variables with an *invalid* value and the first thing any code should do is overwrite this value with something valid. Using an invalid value can not ever be correct.

File: source/base/timer.cc (hunk 7 above)
Comment (masterleinad) [inline source/base/timer.cc]: Previously, `mpi_data` was not used if the `sync_wall_time` flag was not set. In particular,`get_data` and `print_data` didn't return anything useful then. I wanted to fix this and fill this variable in any case. 
`sync_wall_time` just makes sure that `last_lap_time` and `last_lap_cpu_time` (and consequently `cumulative_wall_time` and `cumulative_time`) are set to the maximum value with respect to all processes.
Of course, this change requires a `MPI_Allreduce` call and thus creates overhead also for `sync_wall_time==false`. 
In the end, filling `mpi_data` and `mpi_total_data` with different values on each process in case `sync_wall_time==false` would also be fine with me although it somewhat contradicts the purpose of these variables.

File: doc/news/changes/minor/20170623DanielArndt [hunk 8]
```diff
@@ -0,0 +1,5 @@
+New: The class Timer has the new functions get_total_data(),
+print_total_data, last_wall_time, cpu_time and last_cpu_time()
+for accessing information about the total run and the last lap now.
```
Comment (bangerth) [inline doc/news/changes/minor/20170623DanielArndt]: There are just too many Germanisms in this sentence for me to let it go :-)

Write this as "The Timer class now has new functions Timer::get_total_data(), Timer::print_total_data(), ... to access information about the total run time as well as the last lap."

File: source/base/timer.cc (hunk 2 above)
Comment (bangerth) [inline source/base/timer.cc]: So looking this over again, if the goal is consistency, why not initialize *all* data to a signaling NaN?

**Comment by masterleinad** [comment]
@bangerth I removed the expression `last lap` from the documentation of the `Timer` class and initialize the respective variables with signaling NaNs again.

**Comment by bangerth** [comment]
OK with me then, thanks!

I'll let the others who had questions make the final call.

**Comment by tamiko** [comment]
@kronbichler Some final comments before we merge?

**Comment by kronbichler** [comment]
No, I'm fine with merging. If one does not want synchronization, it is easy enough to do without the MPI_COMM.
